### PR TITLE
Core: Add skip dispose option to ClientApi

### PR DESCRIPTION
**Pull request by Gongreg**
Issue: https://github.com/storybookjs/react-native/issues/11

## What I did

Added an option to skip dispose call inside storiesOf.add() .

Relevant PR:

https://github.com/storybookjs/react-native/pull/30

## How to test

Added tests.

### CODE REVIEW

**Comment by shilman** [comment]
cc @tmeasday

File: lib/client-api/src/client_api.test.ts [hunk 1]
```diff
@@ -548,14 +548,40 @@ describe('preview.client_api', () => {
       expect(Object.values(args.stories).map(v => v.kind)).toEqual(['kind0', 'kind1', 'kind2']);
       expect(getStorybook().map(story => story.kind)).toEqual(['kind1', 'kind2']);
     });
+
+    it('should bind dispose inside add and soriesOf by default', () => {
```
Comment (tmeasday) [inline lib/client-api/src/client_api.test.ts]:
```suggestion
    it('should call `module.hot.dispose` inside add and soriesOf by default', () => {
```

File: lib/client-api/src/client_api.test.ts [hunk 2]
```diff
@@ -548,14 +548,40 @@ describe('preview.client_api', () => {
...
+      const {
+        clientApi: { storiesOf, getStorybook },
+      } = getContext();
+
+      storiesOf('kind', module).add('story', jest.fn());
+
+      expect(module.hot.dispose.calls.length).toEqual(2);
+    });
+
+    it('should not bind dispose inside add when disableAddStoryDispose is true', () => {
```
Comment (tmeasday) [inline lib/client-api/src/client_api.test.ts]:
```suggestion
    it('should not call `module.hot.dispose`  inside add when disableAddStoryDispose is true', () => {
```

**Review by tmeasday — COMMENTED** [review]
@Gongreg I wonder if the option could be called something a bit clearer, maybe `noStoryModuleHotDispose`? Also perhaps we could put a comment in around the option explaining the rationale as we are sure to forget why it is there

**Comment by Gongreg** [comment]
> @Gongreg I wonder if the option could be called something a bit clearer, maybe `noStoryModuleHotDispose`? Also perhaps we could put a comment in around the option explaining the rationale as we are sure to forget why it is there

I am having trouble with thinking of the name. This is not fully disabling hot story dispose. This is disabling one of the two dispose calls.

I is more of `noStoryModuleAddMethodHotDispose` :D

**Comment by tmeasday** [comment]
I guess what I was thinking is we are keeping the dispose for the kind (in `.storiesOf`) but not for the story (in `.add`) -- is that right?

**Comment by Gongreg** [comment]
That’s correct.

**Review by tmeasday — APPROVED** [review]
Haha, WFM.

**Comment by dannyhw** [comment]
Sorry to leave comments on old PR's but these changes are part of the fast refresh fix for react native https://github.com/storybookjs/react-native/pull/30 any chance we could see this in another 5.3.x patch?

**Pull request merged**